src/api/schema/topology.rs

```diff
@@ -0,0 +1,65 @@
+use crate::config::{Config, DataType};
+use async_graphql::{Enum, Object};
+use lazy_static::lazy_static;
+use std::sync::{Arc, RwLock};
+
+#[Enum]
+pub enum SourceOutputType {
+    Any,
+    Log,
+    Metric,
+}
+
+#[derive(Clone)]
+pub struct SourceData {
+    name: String,
+    output_type: DataType,
+}
+
+#[derive(Clone)]
+pub struct Source(SourceData);
+
+lazy_static! {
+    static ref SOURCES: Arc<RwLock<Vec<Source>>> = Arc::new(RwLock::new(vec![]));
+}
+
+#[Object]
+impl Source {
+    /// Source name
+    async fn name(&self) -> String {
+        self.0.name.clone()
+    }
+
+    /// The output type given by the source
+    async fn output_type(&self) -> SourceOutputType {
+        match self.0.output_type {
+            DataType::Metric => SourceOutputType::Metric,
```

Comment:
You can do `impl Into<SourceOutputType> for DataType {}` to get a similar, but more idiomatic API. :)

Reply:
Thanks! I actually used `From<T>` [here](https://github.com/timberio/vector/blob/master/src/api/schema/metrics.rs#L51-L55), and then proceeded to forget that was an option :-) I used `From<T>` to get `Into<T>` for free, although perhaps that happens the other way round too?

Async-graphql has a new [Enum derive](https://async-graphql.github.io/async-graphql/en/define_enum.html), which implements `Into<T>` by specifying an existing enum to map. Unfortunately, it's only in the newer version, which requires a 1.46 toolchain since some related functionality is unstable in 1.44, so we're blocked from upgrading.

In the meantime, this might be a good candidate for an macro to forego the manual matching of two disparate but otherwise identical enums. Or just a good excuse for me to explore macros a bit more!